src/lib/analytics/segmentOneTrustIntegration.ts

```diff
@@ -0,0 +1,224 @@
+      window.OptanonWrapper = getConsentAndLoad
+    })
+  } else {
+    // OneTrust is disabled. Load Segment normally - load all Segment destinations.
+    // @ts-ignore
+    window.analytics.load(SEGMENT_WRITE_KEY)
+  }
+}
+
+async function getConsentAndLoad() {
+  const consent = await getConsent()
+
+  if (consent === PREVIOUS_CONSENT) {
+    // consent didn't change, nothing to do.
+    return
+  }
+
+  PREVIOUS_CONSENT = consent
+
+  // flag each Segment destination on or off, based on consent.
+  const destinationPref = setSegmentDestinationPref(
+    consent,
+    SEGMENT_DESTINATIONS
+  )
+
+  console.log(destinationPref)
+
+  // load Segment and pass on those flags.
+  conditionallyLoadAnalytics({
+    writeKey: SEGMENT_WRITE_KEY,
+    destinations: SEGMENT_DESTINATIONS,
+    destinationPref,
+  })
+}
+
+async function getConsent() {
+  const oneTrustConsent = await getOneTrustConsent()
+
+  if (oneTrustConsent === "") {
+    // failed getting OneTrust consent, return C0001 as we have implicit consent for Strictly Necessary things.
+    return "C0001"
+  } else {
+    return oneTrustConsent
+  }
+}
+
+async function getOneTrustConsent() {
+  let attempts = 0
+  const maxAttempts = 100
+
+  // if OneTrust is not ready, wait 1 second at most.
+  while (!oneTrustReady() && attempts <= maxAttempts) {
+    await delay(10)
+    attempts++
+  }
+
+  if (oneTrustReady()) {
+    // OneTrust stores consent in window.OnetrustActiveGroups.
+    return window.OnetrustActiveGroups
+  } else {
+    return ""
+  }
+}
+
+function oneTrustReady() {
+  // OneTrust is ready when window.OnetrustActiveGroups contains at least C0001.
+  if (
+    typeof window.OnetrustActiveGroups === "string" &&
+    window.OnetrustActiveGroups.split(",").includes("C0001")
+  ) {
+    return true
+  }
+  return false
+}
+
+function delay(n) {
+  return new Promise<void>(done => {
+    setTimeout(() => {
+      done()
+    }, n)
+  })
+}
+
+async function fetchDestinations(writeKey) {
+  let destinations = await fetchDestinationForWriteKey(writeKey)
+
+  destinations = [
+    ...destinations
+      .reduce((map, item) => {
+        if (item.id === "Repeater") return map // remove Segment Repeater destinations
+        map.has(item["id"]) || map.set(item["id"], item)
+        return map
+      }, new Map()) // return object
+      .values(),
+  ]
+
+  return destinations
+}
+
+function setSegmentDestinationPref(consent, destinations) {
+  // map Segment destination id to OneTrust cookie category.
+  const segmentToOneTrust = {
+    AdWords: "C0004",
+    "Amazon S3": "C0001",
+    Appboy: "C0001",
+    "Facebook Pixel": "C0004",
+    "Google Analytics": "C0002",
+    Indicative: "C0002",
+    "Segment.io": "C0001",
+  }
+
+  const consentArray = consent.split(",")
+
+  const basePref = {
+    // default to false for all Segment destinations.
+    // so if we don't set preference for a destination, it will not be loaded.
+    All: false,
```

Comment:
Does this block the unrecognized destination for all visitors or those who'v opted out of some categories?

It seems like a safer default would be `true`, requiring destinations to be treated as "necessary" unless configured otherwise in the mapping above.

Reply:
Correct, with `All: false`, any unrecognized destination (one returned by Segment API but not listed in `segmentToOneTrust` ) will not be loaded.

I wonder how much risk there is with that. This should only be of concern when we come to add a destination on Segment's side. I imagine we will discover early (in staging perhaps) that the destination does not work, and leaern to update `segmentToOneTrust`.

With `All: true` (the default), the destination will work from the get-go. We will likely never update `segmentToOneTrust` (even if there's a playbook). The destination is likely not `Necessary`. We might violate visitor's consent, for a long time.

Reply:
Whatever we choose, we'll have the opposite problem (either integrations not operating as expected or consent not being respected).

This is acceptable as long as we're clear about the burden on developers/analysts.